cpp/src/arrow/record_batch.h

```diff
@@ -23,6 +23,7 @@
 #include <vector>
 
 #include "arrow/compare.h"
+#include "arrow/device.h"
```

Comment:
Can we instead add the necessary declarations to `type_fwd.h`?

Reply:
We would have to move the nested `Device::SyncEvent` outside of the `Device` class to do that since you can't forward declare an inner class like that. It would also require pushing the entire `DeviceAllocationType` declaration to `type_fwd.h` which I'm not sure if we want to do. So I don't think we can do this.

Reply:
I think at least moving `DeviceAllocationType` to `type_fwd.h` would make sense.

As for `Device::SyncEvent`, that's indeed a problem with nested classes (and a good reason to avoid them :-)). That could be fixed by moving `SyncEvent` out of `Device` and adding a compatibility alias.